File: Python codes/stitched_colorbars/stiched_colormap.py

def stiched_colormap(c1, c2, stich_point):
    """
    Creates a modified colormap by stitching together two other colormaps.
    
    Parameters:
        c1 (ndarray): First colormap array (Nx3 or Nx4, where N is the number of colors).
        c2 (ndarray): Second colormap array (Mx3 or Mx4, where M is the number of colors).
        stich_point (float): The point where the two colormaps will be stitched, as a proportion (0.0 to 1.0).

    Returns:
        ListedColormap: A new colormap created by stitching c1 and c2.
    """
    from matplotlib.colors import ListedColormap
    import numpy as np

    # Ensure stitch_point is within valid range
    if not (0.0 < stich_point < 1.0):
        raise ValueError("stich_point must be between 0.0 and 1.0")

    # Set subsampling factors based on the stitch point
    subsample_factor1 = stich_point
    subsample_factor2 = 1 - subsample_factor1

    # Compute new array length and interpolate colors for the first colormap
    new_length1 = round(subsample_factor1 * len(c1))
    new_cmap1 = np.array([np.interp(np.linspace(0, 1, new_length1), np.linspace(0, 1, len(c1)), c1[:, i]) for i in range(c1.shape[1])]).T

    # Compute new array length and interpolate colors for the second colormap
    new_length2 = round(subsample_factor2 * len(c2))
    new_cmap2 = np.array([np.interp(np.linspace(0, 1, new_length2), np.linspace(0, 1, len(c2)), c2[:, i]) for i in range(c2.shape[1])]).T

    # Combine the two parts into one array
    combined_colors = np.vstack((new_cmap1, new_cmap2))

    # Create a new ListedColormap with the combined colors
    cdis = ListedColormap(combined_colors)
    return cdis

File: matlab/colormaps/stiched_colormap.m
function [cdis] = stiched_colormap(cmap1, cmap2, stich_point)
    % % Creates a modified colormap by stitching together two other colormaps.
    % % cmap1: name of first colormap
    % % cmap2: name of second colormap
    % % stich_point: the point where the two colormaps will be stitched (0-100)
    
    % Import necessary libraries
    %import matlab.graphics.*
    
    % Set subsampling factor
    subsample_factor= stich_point/100;
    
    %%% Get colors from first colormap up to stitch point %%%
    
    % Compute new array length
    new_length1 = round(subsample_factor * size(cmap1, 1));
    % Create new array by linearly interpolating between original values
    new_cmap1 = interp1(1:size(cmap1, 1), cmap1, linspace(1, size(cmap1, 1), new_length1), 'linear');
    pixels1=size(new_cmap1, 1);
    
    
    %%% Get colors from second colormap from stitch point to end %%%
    
    % Compute new array length
    subsample_factor2=1-subsample_factor;
    new_length2 = round(subsample_factor2 * size(cmap2, 1));
    
    % Create new array by linearly interpolating between original values
    new_cmap2 = interp1(1:size(cmap2, 1), cmap2, linspace(1, size(cmap2, 1), new_length2), 'linear');
    pixels2=size(new_cmap2, 1);
    disp(['Check total pixels: ' num2str(pixels1+pixels2)]);
    
    cdis=cmap1;
    cdis(1:length(new_cmap1),:,:)=new_cmap1;
    cdis(pixels1+1:pixels1+pixels2,:,:)=new_cmap2;
end

